In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import plotly
import plotly.figure_factory as ff
import plotly.graph_objs as go
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import SGDClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler
from plotly.offline import download_plotlyjs, init_notebook_mode, plot, iplot
init_notebook_mode(connected=True)

In [3]:
data = pd.read_csv('task_b.csv')
data=data.iloc[:,1:]

In [4]:
data.head()

,f1,f2,f3,y
0,-195.871045,-14843.084171,5.532140,1.0
1,-1217.183964,-4068.124621,4.416082,1.0
2,9.138451,4413.412028,0.425317,0.0
3,363.824242,15474.760647,1.094119,0.0
4,-768.812047,-7963.932192,1.870536,0.0


In [5]:
data.corr()['y']

f1    0.067172
f2   -0.017944
f3    0.839060
y     1.000000
Name: y, dtype: float64

In [6]:
data.std()

f1      488.195035
f2    10403.417325
f3        2.926662
y         0.501255
dtype: float64

In [7]:
data.mean()

f1      10.180031
f2    1299.986739
f3       5.001840
y        0.500000
dtype: float64

In [8]:
X=data[['f1','f2','f3']].values
Y=data['y'].values
print(X.shape)
print(Y.shape)

(200, 3)
(200,)


# What if our features are with different variance 

<pre>
* <b>As part of this task you will observe how linear models work in case of data having feautres with different variance</b>
* <b>from the output of the above cells you can observe that var(F2)>>var(F1)>>Var(F3)</b>

> <b>Task1</b>:
    1. Apply Logistic regression(SGDClassifier with logloss) on 'data' and check the feature importance
    2. Apply SVM(SGDClassifier with hinge) on 'data' and check the feature importance

> <b>Task2</b>:
    1. Apply Logistic regression(SGDClassifier with logloss) on 'data' after standardization 
       i.e standardization(data, column wise): (column-mean(column))/std(column) and check the feature importance
    2. Apply SVM(SGDClassifier with hinge) on 'data' after standardization 
       i.e standardization(data, column wise): (column-mean(column))/std(column) and check the feature importance

</pre>

<h3><font color='blue'> Make sure you write the observations for each task, why a particular feautre got more importance than others</font></h3>

# Applying logistic regression to the data 

In [9]:
clf = SGDClassifier(loss="log", penalty="l2")
clf.fit(X, Y)
train = clf.fit(X,Y)
yhat = clf.predict(X)
print(train.coef_[0],train.intercept_[0])
#draw_line(train.coef_[0],train.intercept_[0],np.min(X),np.max(X))
#plt.scatter(X[:,0],X[:,1],c=Y)

[ 3855.61449376 24007.99876578 10341.96821606] 81.22274044416113


# Observation

After applying the logistic regression the coeffiecient value is [3855.61449376 24007.99876578 10341.96821606]
and intercept is 81.2227

# Applying SVM to the dataset

In [10]:
clf = SGDClassifier(loss="hinge", penalty="l2")
clf.fit(X, Y)
train = clf.fit(X,Y)
yhat = clf.predict(X)
print(train.coef_[0],train.intercept_[0])
print(train)
#draw_line(train.coef_[0],train.intercept_[0],np.min(X),np.max(X))
#plt.scatter(X[:,0],X[:,1],c=Y)

[ 9785.50801039  1133.21700643 10435.23045859] 318.5905864781299


# Observation 

After applying the logistic regression the coeffiecient value is [ 9785.50801039  1133.21700643 10435.23045859]
and intercept is 318.5905

# Standardized the data

In [11]:
df_stand = pd.DataFrame(columns=['f1_std', 'f2_std','f3_std']).astype('float64')

In [12]:
print(df_stand)

Empty DataFrame
Columns: [f1_std, f2_std, f3_std]
Index: []


In [13]:
f1_std = []
f2_std = []
f3_std = []
for value in data["f1"]:
    value = ((value - data['f1'].mean())/(data['f1'].std()))
    f1_std.append(value)
for value in data["f2"]:
    value = ((value - data['f2'].mean())/(data['f2'].std()))
    f2_std.append(value)
for value in data["f3"]:
    value = ((value - data['f3'].mean())/(data['f3'].std()))
    f3_std.append(value)            

In [14]:
df_stand["f1_std"] = f1_std
df_stand["f2_std"] = f2_std
df_stand["f3_std"] = f3_std

In [15]:
print(df_stand.head())

     f1_std    f2_std    f3_std
0 -0.422067 -1.551708  0.181196
1 -2.514085 -0.515995 -0.200146
2 -0.002134  0.299269 -1.563735
3  0.724391  1.362511 -1.335214
4 -1.595658 -0.890469 -1.069923


In [16]:
X_std=df_stand[['f1_std','f2_std','f3_std']].values
print(X_std.shape)
print(Y.shape)

(200, 3)
(200,)


# Applying LR on standradized data

In [17]:
clf = SGDClassifier(loss="log", penalty="l2")
clf.fit(X_std, Y)
train = clf.fit(X_std,Y)
yhat = clf.predict(X_std)
print(train.coef_[0],train.intercept_[0])
#draw_line(train.coef_[0],train.intercept_[0],np.min(X_std),np.max(X_std))
#plt.scatter(X_std[:,0],X_std[:,1],c=Y)

[-0.19066245  0.3667424  10.5107287 ] -1.876270597276803


# Observation

The coefficient on standardized data [-0.19066245  0.3667424  10.5107287 ]  and intercept is -1.876270597276803

In [18]:
clf = SGDClassifier(loss="hinge", penalty="l2")
clf.fit(X_std, Y)
train = clf.fit(X_std,Y)
yhat = clf.predict(X_std)
print(train.coef_[0],train.intercept_[0])

[-1.19308516  4.91035635 11.81005766] 1.2996165329936604


# Observation

The coefficient on standardized data [-1.19308516  4.91035635 11.81005766] and inntercept is 1.2996165329936604

After standarization the value of coffecient and intercept changed rapidly and value of the feature (f1) is coming as a negative and 
the intercept value is coming as negative 